Removed netCDF4 package from integration tests of fixes

schlunma opened the pull request:
<!--
    Thank you for contributing to our project!
-->

This PR removes the use of the package `netCDF4` in test for fixes to avoid segmentation faults.

<!--
    Please describe your changes here, especially focusing on why this PR makes
    ESMValCore better and what problem it solves.

    Before you start, please read our [contribution guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html).

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->

-   Addresses #644 (I don't know if it closes this issue completely)

***

## Before you get started

-   [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## Checklist

-   [x] PR has a descriptive title for the [changelog](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review)
-   [x] Code follows the [style guide](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#code-style)
-   [x] [Documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#documentation) is available for new functionality
-   [x] YAML files pass [`pre-commit`](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#pre-commit) or [`yamllint`](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/community/introduction.html#yaml) checks
-   [x] [Circle/CI tests pass](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review)
-   [x] [Codacy code quality checks pass](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review)
-   [x] [Documentation builds successfully](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#branches-pull-requests-and-code-review) on readthedocs
-   [x] [Unit tests](https://docs.esmvaltool.org/projects/esmvalcore/projects/esmvalcore/en/latest/contributing.html#contributing-to-the-esmvalcore-package) are available

***

To help with the number pull requests:

-   🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

valeriupredoi commented:
dat was superqwik, cheers Manu! I am having a look and stress test now :+1:

valeriupredoi commented:
good job with the PR @schlunma :+1: :beer: 

- the PR helps with #644 and makes things cleaner but it doesn't solve #644 because of this [comment](https://github.com/ESMValGroup/ESMValCore/issues/644#issuecomment-758775667) and this [comment](https://github.com/ESMValGroup/ESMValCore/issues/644#issuecomment-758784798) although the PR does narrow down the problem and makes sure there are no stray thread-unsafe issues from netCDF4 in the tests, since yes, netCDF4 is thread-unsafe
- minor comment: there is one more import of `netCDF4` in `tests/integration/cmor/_fixes/sample_data/create_sample_data.py` - do we want that changed too?

I reckon this is good to have and we need to move on understanding the iris-dask animosity from #644 but that's not the scope of this PR :+1:

schlunma commented:
Cheers V, thanks for reviewing! :beer: 

>     * minor comment: there is one more import of `netCDF4` in `tests/integration/cmor/_fixes/sample_data/create_sample_data.py` - do we want that changed too?

Nope, we can't get rid of that since those (faulty) data cannot be produced with `iris`. However, this should not matter since this script is not used during the test. I only used it to create the sample data and I thought it would be good to keep it in the repo :+1:

valeriupredoi commented:
> Cheers V, thanks for reviewing! beer
> 
> > ```
> > * minor comment: there is one more import of `netCDF4` in `tests/integration/cmor/_fixes/sample_data/create_sample_data.py` - do we want that changed too?
> > ```
> 
> Nope, we can't get rid of that since those (faulty) data cannot be produced with `iris`. However, this should not matter since this script is not used during the test. I only used it to create the sample data and I thought it would be good to keep it in the repo +1

excellent! Let's wait until tomorrow morning see if any of the other folks want to have a look, if no answer, fire it to master, bud! :beer:

stefsmeets commented on tests/integration/cmor/_fixes/cmip6/test_cesm2.py:
```diff
@@ -7,82 +7,13 @@
-        dataset.variables['cl'].units = '%'
-
-    return nc_path
+SAMPLE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                           os.pardir, 'sample_data')
```
Would be good if we use pathlib where possible!

```suggestion
from pathlib import Path

SAMPLE_PATH = Path(__file__).absolute().parents[1] / 'sample_data'
```

stefsmeets commented on tests/integration/cmor/_fixes/cmip6/test_cesm2.py:
```diff
@@ -128,9 +59,10 @@ def test_get_cl_fix():
+def test_cl_fix_file(mock_get_filepath, tmp_path):
     """Test ``fix_file`` for ``cl``."""
-    cubes = iris.load(cl_file)
+    nc_path = os.path.join(SAMPLE_PATH, 'cesm2_cl.nc')
```
```suggestion
    nc_path = SAMPLE_PATH / 'cesm2_cl.nc'
```

stefsmeets commented on tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py:
```diff
@@ -6,48 +6,13 @@
-
-    dataset.close()
-    return nc_path
+SAMPLE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                           os.pardir, 'sample_data')
```
```suggestion
from pathlib import Path

SAMPLE_PATH = Path(__file__).absolute().parents[1] / 'sample_data'
```

stefsmeets replied:
This statement is at the top of every file. Would it be an idea to create a fixture that defines the sample path? Or some other way to single-source this (e.g. via an import or in conftest.py)?

stefsmeets commented on tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py:
```diff
@@ -61,25 +26,25 @@ def test_get_cl_fix():
-def test_cl_fix_file(mock_get_filepath, cl_file, tmp_path):
+def test_cl_fix_file(mock_get_filepath, tmp_path):
     """Test ``fix_file`` for ``cl``."""
+    nc_path = os.path.join(SAMPLE_PATH, 'cesm2_waccm_cl.nc')
```
```suggestion
    nc_path = SAMPLE_PATH / 'cesm2_waccm_cl.nc'
```

stefsmeets reviewed (changes requested):
Hi @schlunma , nice work, looks much cleaner! 

1. The definition of the `SAMPLE_PATH` variable is repeated at the top of every file. It would be nice if this could be single-sourced. Also, would it be possible for you to change the definition from using `os.path` to `pathlib.Path` while at it (I already made a few starts)? This will be much easier to work with and maintain.

2. Could you put the scripts to generate the `.nc` files somewhere (maybe directly in the `sample_data` directory) so that the data can be regenerated or updated if needed?

3. Finally, we already have a set of tests with sample data here: https://github.com/ESMValGroup/ESMValCore/tree/master/tests/sample_data/ . Could we call your directory `test_data` to avoid a name clash?

valeriupredoi commented:
@stefsmeets very good idea about adding instructions how to create those standard test files, cheers! About `Pathlib` - you know me, am impartial :grin: 
More generally speaking, I need to put together a small reproducible test that replicates the issue related to data realization when run on multiple threads and the SegFault we get so we can ship it to the iris folk - will do this now :+1:

valeriupredoi commented:
let's get this one done and merged, I'd like to continue the fixing trials by reducing the size of tests, am fairly sure it's the size of the tests that's causing the SegFaults as I say in this [comment](https://github.com/ESMValGroup/ESMValCore/issues/644#issuecomment-759632105) - Manu, if you don't have time I can tend to Stef's comments :+1:

schlunma commented:
I already addressed them and will push them after the meeting 👍

valeriupredoi commented:
> I already addressed them and will push them after the meeting +1

total boss! :beer:

schlunma commented:
Thanks for the review @stefsmeets !

>     1. The definition of the `SAMPLE_PATH` variable is repeated at the top of every file. It would be nice if this could be single-sourced. Also, would it be possible for you to change the definition from using `os.path` to `pathlib.Path` while at it (I already made a few starts)? This will be much easier to work with and maintain.

I changed `os.path` to `pathlib` and moved the sample path to a fixture in `conftest.py`. I'm quite happy with `pathlib`, although it's a pity that `iris` can't handle `Path` objects at the moment.

>     2. Could you put the scripts to generate the `.nc` files somewhere (maybe directly in the `sample_data` directory) so that the data can be regenerated or updated if needed?

The script is already there: [tests/integration/cmor/_fixes/test_data/create_test_data.py](https://github.com/ESMValGroup/ESMValCore/blob/c12f3982b7303f2f5589da8f45d8112c7f5512dd/tests/integration/cmor/_fixes/test_data/create_test_data.py)

>     3. Finally, we already have a set of tests with sample data here: https://github.com/ESMValGroup/ESMValCore/tree/master/tests/sample_data/ . Could we call your directory `test_data` to avoid a name clash?

Renamed to `test_data`.

valeriupredoi commented:
good job @schlunma - this change reduces the statistical rate of SegFaults from ~16% to 5/80~6% (16% from this [comment](https://github.com/ESMValGroup/ESMValCore/issues/644#issuecomment-758609069))

stefsmeets reviewed (approved):
Cheers @schlunma , I must have missed the script to generate the data on my first pass (I think the diff was probably too big for it to show 😅). Nice work, hope this helps reduce the number of seg faults!

valeriupredoi commented:
cheers @schlunma and @stefsmeets :beer:
